### PR TITLE
Improve LUA asset dependency processing

### PR DESCRIPTION
**Pull request by vmednonogov**
## What does this PR do?

Improve LUA asset dependency processing

## How was this PR tested?

Tested locally

### CODE REVIEW

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 1]
```diff
@@ -43,12 +43,29 @@ namespace LuaBuilder
                 {
                     AZStd::string watchFolder;
                     AZ::Data::AssetInfo assetInfo;
-                    AZ::IO::Path path(dependency.m_dependencyPath);
-                    bool isLuaDependency = !path.HasExtension() || path.Extension() == ".lua" || path.Extension() == ".luac";
-                    auto sourcePath = path.ReplaceExtension(".lua");
+                    // carbonated begin
```
Comment (alexeykostin) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Please add `#if defined(CARBONATED)`

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 2]
```diff
@@ -43,12 +43,29 @@ namespace LuaBuilder
...
+                    AZ::IO::Path path(dependency.m_dependencyPath.starts_with("/")
+                        ? dependency.m_dependencyPath.substr(1) : dependency.m_dependencyPath); // path shoul not start from a path delimiter
 
-                    if (assetSystem->GetSourceInfoBySourcePath(sourcePath.c_str(), assetInfo, watchFolder)
-                        && assetInfo.m_assetId.IsValid())
+                    if (!path.HasExtension() || path.Extension() == ".luac") // assume that a file witout extension is a lua file, ".luac" is generated from ".lua"
```
Comment (alexeykostin) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: You can get once and reuse extension:

`const auto pathExtension = path.Extension();`

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 3]
```diff
@@ -43,12 +43,29 @@ namespace LuaBuilder
                 {
                     AZStd::string watchFolder;
                     AZ::Data::AssetInfo assetInfo;
-                    AZ::IO::Path path(dependency.m_dependencyPath);
-                    bool isLuaDependency = !path.HasExtension() || path.Extension() == ".lua" || path.Extension() == ".luac";
-                    auto sourcePath = path.ReplaceExtension(".lua");
+                    // carbonated begin
+                    AZ::IO::Path path(dependency.m_dependencyPath.starts_with("/")
+                        ? dependency.m_dependencyPath.substr(1) : dependency.m_dependencyPath); // path shoul not start from a path delimiter
```
Comment (alexeykostin) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: shoul not -> should not

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 2 above)
Comment (alexeykostin) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: witout -> without

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 4]
```diff
@@ -43,12 +43,29 @@ namespace LuaBuilder
...
+                    }
+                    else if (path.Extension() == ".dynamicslice") // ".dynamicslice" is generated from ".slice"
+                    {
+                        path.ReplaceExtension(".slice");
+                    }
+                    else if (path.Extension() == ".spawnable") // ".spawnable" is generated from ".prefab"
+                    {
+                        path.ReplaceExtension(".prefab");
+                    }
+                    // todo: replace ".xxx.streamingimage" --> ".xxx" (where xxx is png, jpg, tif, ...), it requires to determine an original image path
```
Comment (carb-lloyd) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Create a map that you can iterate, instead of if/else/if/else

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 5]
```diff
@@ -57,17 +74,11 @@ namespace LuaBuilder
...
-                        AZ_Error("LuaBuilder", false, "Did not find dependency %s referenced by script.", dependency.m_dependencyPath.c_str());
-                    }
                     else
                     {
-                        AZ_Error("LuaBuilder", false, "%s referenced by script does not appear to be a lua file.\n"
-                            "References to assets should be handled using Property slots and AssetIds to ensure proper dependency tracking.\n"
-                            "This file will not be tracked as a dependency.",
-                            dependency.m_dependencyPath.c_str());
+                        AZ_Error("LuaBuilder", false, "Invalid asset %s referenced by script.", dependency.m_dependencyPath.c_str());
                     }
```
Comment (carb-lloyd) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: This may be a false positive for strings that are just plain strings.  Don't error out, make it a warning instead (or maybe even and info line).  "String '%s' is not a valid asset, will not add to dependency list" or something like that.  We don't want errors from false positives.

**Review by carb-lloyd — CHANGES_REQUESTED** [review]
Concurred with @alexeykostin re: the #if defined(CARBONATED) guards.

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 3 above)
Comment (aefimov3) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Sometimes there might be a backslash, typically there is not, but this is a lua strings, so there can be any characters. I believe engine process a patch with back slashes OK, it changes \ to / at some moment.

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 6]
```diff
@@ -57,17 +74,11 @@ namespace LuaBuilder
...
-                    {
-                        AZ_Error("LuaBuilder", false, "Did not find dependency %s referenced by script.", dependency.m_dependencyPath.c_str());
-                    }
                     else
                     {
-                        AZ_Error("LuaBuilder", false, "%s referenced by script does not appear to be a lua file.\n"
-                            "References to assets should be handled using Property slots and AssetIds to ensure proper dependency tracking.\n"
-                            "This file will not be tracked as a dependency.",
-                            dependency.m_dependencyPath.c_str());
+                        AZ_Error("LuaBuilder", false, "Invalid asset %s referenced by script.", dependency.m_dependencyPath.c_str());
```
Comment (aefimov3) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: It makes sense to mention that the asset is not recognized, there might be an asset type that is not supported by the asset map yet. So there should be a hint to developer on fixing the issue.

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 7]
```diff
@@ -57,17 +74,11 @@ namespace LuaBuilder
...
                     }
+                    // carbonated end
```
Comment (aefimov3) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: You did not change the later code where every job that processes assets in the list gets a lua-related string:
            for (auto& dependentAsset : dependentAssets)
            {
                AssetBuilderSDK::JobDependency jobDependency;
                jobDependency.m_sourceFile.m_sourceFileDependencyUUID = dependentAsset.GetId().m_guid;
                jobDependency.m_jobKey = "Lua Compile";

**Review by aefimov3 — CHANGES_REQUESTED** [review]

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 1 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Added

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 2 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: This variable was removed at all, due to now we use a map for extension replacement

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 3 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: 1) updated
2) updated, removed both kind of slashes in the first position. Yes, I have checked, AP works fine with both kind of slashes inside a filepath. It can't work with the slash in the first path position (o3de specific)

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 2 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Updated

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 4 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Replaced to map

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 5 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Replaced to string "String '%s' is not a valid asset, will not add to dependency list" and changed to warning

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 6 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Added a clarification line

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 7 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Looks like jobKey for dependency is skipped always, when dependency jobs are analyzed (see AssetProcessorManager::UpdateSourceFileDependenciesDatabase for example) and later AP sets an appropriate jobKey for any resource type from dependency or remove duplicated dependencies.
 
I can't catch no one case in AP at the page "Jobs" in column "Job Key"  when "Lua Compile" appears for any kind of assets different with LUA. "Job Key" column are always displayed properly. Need further investigation.

In any case, what is the Job Key:
```
            //! Job Key is arbitrarily defined by the builder.  Used to differentiate between different jobs emitted for the same input file, for the same platform, for the same builder.
            //! for example, you might want to split a particularly complicated and time consuming job into multiple sub-jobs.  In which case they'd all have the same input file,
            //! the same platform, the same builder UUID (since its the UUID of the builder itself)
            //! but would have different job keys.
```

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 8]
```diff
@@ -33,6 +33,20 @@ namespace LuaBuilder
...
     {
+// carbonated begin
+#if defined(CARBONATED)
+        static const AZStd::unordered_map<AZStd::string, AZStd::string> AssetExtensionReplacementMap =
+        {
+            {"",                 ".lua"},     // assume that a file without extension maybe a lua file
+            {".luac",            ".lua"},     // ".luac" is generated from ".lua"
+            {".dynamicslice",    ".slice"},   // ".dynamicslice" is generated from ".slice"
+            {".spawnable",       ".prefab"},  // ".spawnable" is generated from ".prefab"
+            {".streamingimage",  ""},         // ".xxx.streamingimage" is generated from ".xxx" (where xxx is png, jpg, tif, ...)
```
Comment (vmednonogov) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp:42]: Verified ^^^^
Added line ` ftueImageName = "/dlc/abilities/models/ability_aim_diff.tga.streamingimage"`
into one ftue LUA file and it was processed properly

**Review by aefimov3 — APPROVED** [review]

**Review by alexeykostin — APPROVED** [review]

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 8 above)
Comment (carb-lloyd) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp:42]: Remove the // carbonated begin/end :/

Unnecessary given the define.

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 9]
```diff
@@ -43,6 +57,36 @@ namespace LuaBuilder
                 {
                     AZStd::string watchFolder;
                     AZ::Data::AssetInfo assetInfo;
+// carbonated begin
+#if defined(CARBONATED)
```
Comment (carb-lloyd) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: likewise.

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 10]
```diff
@@ -33,6 +33,20 @@ namespace LuaBuilder
 {
     namespace
     {
+// carbonated begin
+#if defined(CARBONATED)
+        static const AZStd::unordered_map<AZStd::string, AZStd::string> AssetExtensionReplacementMap =
+        {
+            {"",                 ".lua"},     // assume that a file without extension maybe a lua file
```
Comment (carb-lloyd) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Let's not make this assumption.  The AP doesn't process files without an extension as LUA files, and neither should we.  If a dev referenced a lua file without extension, then that is their error to fix.  

TL;DR:  remove the "" -> ".lua" mapping

**Review by carb-lloyd — CHANGES_REQUESTED** [review]

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 9 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Removed the // carbonated begin/end (here and above)

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 10 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Removed this rule.
It allows to skip any file path without extension.

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 8 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp:42]: Removed

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp [hunk 11]
```diff
@@ -43,6 +54,42 @@ namespace LuaBuilder
...
+                        if (it != AssetExtensionReplacementMap.end())
+                        {
+                            path.ReplaceExtension(it->second.c_str());
+                        }
+
+                        if (!assetSystem->GetSourceInfoBySourcePath(path.c_str(), assetInfo, watchFolder))
+                        {
+                            AZ_Warning("LuaBuilder", false, "Did not find dependency '%s' referenced by script (or it is not an asset path).", dependency.m_dependencyPath.c_str());
+                        }
+                        else if (assetInfo.m_assetId.IsValid())
```
Comment (carb-lloyd) [inline Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Minor change here.  Clean up as follows:
```
bool success = assetSystem->GetSourceinfoBySourcePath(..)
if(success)
{
    if(assetInfo.. IsValid())
         ... 
    else
         ...
}
else 
   ...
```

It's easier to read - I'm not a fan of assetInfo being set (if successful) in the conditional of the precedening"if" before it.

**Review by carb-lloyd — CHANGES_REQUESTED** [review]

File: Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp (hunk 11 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Builders/LuaBuilder/LuaBuilderWorker.cpp]: Updated as described above ^^^

**Review by carb-lloyd — APPROVED** [review]
